File: one.py
import snntorch as snn
import torch
import itertools

# Part 1
# ------------------------------------------------------------------------------

# def leaky_integrate_neuron(U, time_step=1e-3, I=0, R=5e7, C=1e-10):
#   tau = R*C
#   U = U + (time_step/tau)*(-U + I*R)
#   return U

# num_steps = 100
# U = 0.9
# U_trace = []  # keeps a record of U for plotting

# for step in range(num_steps):
#   U_trace.append(U)
#   U = leaky_integrate_neuron(U)  # solve next step of U

# for i,v in enumerate(U_trace):
#   print(f"{i}: {v:.8f}")

# Part 2
# ------------------------------------------------------------------------------

time_step = 1e-3
R = 5
C = 1e-3

# leaky integrate and fire neuron, tau=5e-3
lif1 = snn.Lapicque(R=R, C=C, time_step=time_step)

# # Initialize membrane, input, and output
# mem = torch.ones(1) * 0.9  # U=0.9 at t=0
# cur_in = torch.zeros(num_steps, 1)  # I=0 for all t
# spk_out = torch.zeros(1)  # initialize output spikes

# # A list to store a recording of membrane potential
# mem_rec = [mem]

# # pass updated value of mem and cur_in[step]=0 at every time step
# for step in range(num_steps):
#   spk_out, mem = lif1(cur_in[step], mem)

#   # Store recordings of membrane potential
#   mem_rec.append(mem)

# # convert the list of tensors into one tensor
# mem_rec = torch.stack(mem_rec)

# for i,v in enumerate(list(itertools.chain(*mem_rec.tolist()))):
#   print(f"{i}: {v:.8f}")

# Part 3
# ------------------------------------------------------------------------------

# # Initialize input current pulse
# cur_in = torch.cat((torch.zeros(10, 1), torch.ones(190, 1)*0.1), 0)  # input current turns on at t=10

# # Initialize membrane, output and recordings
# mem = torch.zeros(1)  # membrane potential of 0 at t=0
# spk_out = torch.zeros(1)  # neuron needs somewhere to sequentially dump its output spikes
# mem_rec = [mem]
# spiked = []
# num_steps = 200

# # pass updated value of mem and cur_in[step] at every time step
# for step in range(num_steps):
#   spk_out, mem = lif1(cur_in[step], mem)
#   spiked.append(spk_out.item())
#   mem_rec.append(mem)

# # crunch -list- of tensors into one tensor
# mem_rec = torch.stack(mem_rec)

# for i,v in enumerate(list(itertools.chain(*mem_rec.tolist()))):
#   print(f"{i}: {v:.8f}")
# print(f"{spiked}")

# print(f"The calculated value of input pulse [A] x resistance [Ω] is: {cur_in[11]*lif1.R} V")
# print(f"The simulated value of steady-state membrane potential is: {mem_rec[200][0]} V")

# Part 4
# ------------------------------------------------------------------------------

num_steps = 200

# # Initialize current pulse, membrane and outputs
# cur_in1 = torch.cat((torch.zeros(10, 1), torch.ones(20, 1)*(0.1), torch.zeros(170, 1)), 0)  # input turns on at t=10, off at t=30
# mem = torch.zeros(1)
# spk_out = torch.zeros(1)
# mem_rec1 = [mem]


# # neuron simulation
# for step in range(num_steps):
#   spk_out, mem = lif1(cur_in1[step], mem)
#   mem_rec1.append(mem)
# mem_rec1 = torch.stack(mem_rec1)

# for i,v in enumerate(list(itertools.chain(*mem_rec1.tolist()))):
#   print(f"{i}: {v:.8f}")

# # Increase amplitude of current pulse; quarter the time.
# cur_in3 = torch.cat((torch.zeros(10, 1), torch.ones(5, 1)*0.147, torch.zeros(185, 1)), 0)  # input turns on at t=10, off at t=15
# mem = torch.zeros(1)
# spk_out = torch.zeros(1)
# mem_rec3 = [mem]

# # neuron simulation
# for step in range(num_steps):
#   spk_out, mem = lif1(cur_in3[step], mem)
#   mem_rec3.append(mem)
# mem_rec3 = torch.stack(mem_rec3)

# for i,v in enumerate(list(itertools.chain(*mem_rec3.tolist()))):
#   print(f"{i}: {v:.8f}")

# Part 5
# ------------------------------------------------------------------------------

# LIF w/Reset mechanism
def leaky_integrate_and_fire(mem, cur=0, threshold=1, time_step=1e-3, R=5.1, C=5e-3):
  tau_mem = R*C
  spk = (mem > threshold)
  mem = mem + (time_step/tau_mem)*(-mem + cur*R) - spk*threshold  # every time spk=1, subtract the threhsold
  return mem, spk

# Small step current input
cur_in = torch.cat((torch.zeros(10), torch.ones(190)*0.2), 0)
mem = torch.zeros(1)
mem_rec = []
spk_rec = []

# neuron simulation
for step in range(num_steps):
  mem, spk = leaky_integrate_and_fire(mem, cur_in[step])
  mem_rec.append(mem)
  spk_rec.append(spk)

# convert lists to tensors
mem_rec = torch.stack(mem_rec)
spk_rec = torch.stack(spk_rec)
for i,v in enumerate(list(itertools.chain(*mem_rec.tolist()))):
  print(f"{i}: {v:.8f}")
for i,v in enumerate(list(itertools.chain(*spk_rec.tolist()))):
  print(f"{i}: {v:.8f}")

File: src/bin/one.rs
use std::iter::repeat;
use std::time::Duration;

fn main() {
    println!("Hello, world!");

    // Part 1
    // -------------------------------------------------------------------------

    // - `u`: The current membrane potential.
    // - `time_step`: The amount of time to step forward.
    // - `i`: Input current (e.g. incoming spike).
    // - `r`: Resistance (Ω).
    // - `c`: Capacative capacity.
    // fn leaky_integrate_neuron(u: f64, time_step: Duration, i: f64, r: f64, c: f64) -> f64 {
    //     let tau = r * c;
    //     return u + (time_step.as_secs_f64() / tau) * (i * r - u);
    // }

    // let num_steps = 100;
    // let potentials = (0..num_steps).scan(0.9f64, |a,_ |{
    //     let b = *a;
    //     *a = leaky_integrate_neuron(
    //         *a,
    //         Duration::from_secs_f64(1e-3_f64),
    //         0f64,
    //         5e7_f64,
    //         1e-10_f64,
    //     );
    //     Some(b)
    // }).collect::<Vec<_>>();
    // for (i,p) in potentials.iter().enumerate() {
    //     println!("{i}: {p:.8}");
    // }

    // Part 2
    // -------------------------------------------------------------------------

    // struct Lif {
    //     r: f64,
    //     c: f64,
    //     time_step: Duration,
    // }
    // impl Lif {
    //     fn step(&self, i: f64, u: f64) -> (bool, f64) {
    //         let membrane_potential = leaky_integrate_neuron(u, self.time_step, i, self.r, self.c);
    //         // Default `threshold` on `snn.Lapicque` is `1`.
    //         (membrane_potential > 1f64, membrane_potential)
    //     }
    // }

    // let lif1 = Lif{ r: 5f64, c: 1e-3_f64, time_step: Duration::from_secs_f64(1e-3_f64)};
    // let mut mem = 0.9;
    // let cur_in = (0..num_steps).map(|_| 0f64).collect::<Vec<_>>();
    // let mut mem_rec = vec![mem];
    // for step in 0..num_steps {
    //     let (_spk_out, new_mem) = lif1.step(cur_in[step], mem);
    //     mem = new_mem;
    //     mem_rec.push(mem);
    // }
    // for (i, p) in mem_rec.iter().enumerate() {
    //     println!("{i}: {p:.8}");
    // }

    // Part 3
    // -------------------------------------------------------------------------

    // let mut mem = 0f64;
    // let cur_in = (0..10).map(|_| 0f64).chain((0..190).map(|_|0.1f64)).collect::<Vec<_>>();
    // let mut mem_rec = vec![mem];
    // let mut spiked = Vec::new();
    // let num_steps = 200;
    // for step in 0..num_steps {
    //     let (spk_out, new_mem) = lif1.step(cur_in[step], mem);
    //     mem = new_mem;
    //     spiked.push(spk_out);
    //     mem_rec.push(mem);
    // }
    // for (i, p) in mem_rec.iter().enumerate() {
    //     println!("{i}: {p:.8}");
    // }
    // println!("{spiked:?}");
    // println!("The calculated value of input pulse [A] x resistance [Ω] is: {} V",cur_in[11]*lif1.r);
    // println!("The simulated value of steady-state membrane potential is: {} V",mem_rec[200]);

    // Part 4
    // -------------------------------------------------------------------------

    // let num_steps = 200;

    // let mut mem = 0f64;
    // let cur_in = repeat(0f64).take(10)
    //     .chain(repeat(0.1f64).take(20))
    //     .chain(repeat(0f64).take(170))
    //     .collect::<Vec<_>>();
    // let mut mem_rec = vec![mem];
    // for step in 0..num_steps {
    //     let (_spk_out, new_mem) = lif1.step(cur_in[step], mem);
    //     mem = new_mem;
    //     mem_rec.push(mem);
    // }
    // for (i, p) in mem_rec.iter().enumerate() {
    //     println!("{i}: {p:.8}");
    // }

    // let mut mem = 0f64;
    // let cur_in = repeat(0f64).take(10)
    //     .chain(repeat(0.147f64).take(5))
    //     .chain(repeat(0f64).take(185))
    //     .collect::<Vec<_>>();
    // let mut mem_rec = vec![mem];
    // let num_steps = 200;
    // for step in 0..num_steps {
    //     let (_spk_out, new_mem) = lif1.step(cur_in[step], mem);
    //     mem = new_mem;
    //     mem_rec.push(mem);
    // }
    // for (i, p) in mem_rec.iter().enumerate() {
    //     println!("{i}: {p:.8}");
    // }

    // Part 5
    // -------------------------------------------------------------------------

    fn leaky_integrate_neuron(
        mem: f64,
        cur: f64,
        threshold: f64,
        time_step: Duration,
        r: f64,
        c: f64,
    ) -> (f64, bool) {
        let tau = r * c;
        let spk = mem > threshold;
        let new_mem = mem + (time_step.as_secs_f64() / tau) * (cur * r - mem)
            - (spk as u8 as f64 * threshold);
        return (new_mem, spk);
    }

    struct Lif {
        r: f64,
        c: f64,
        time_step: Duration,
        threshold: f64,
    }
    impl Lif {
        fn new(r: f64, c: f64, time_step: Duration) -> Self {
            Self {
                r,
                c,
                time_step,
                threshold: 1f64,
            }
        }
        fn step(&self, mem: f64, cur: f64) -> (f64, bool) {
            leaky_integrate_neuron(mem, cur, self.threshold, self.time_step, self.r, self.c)
        }
    }

    let lif1 = Lif::new(5.1f64, 5e-3_f64, Duration::from_secs_f64(1e-3_f64));

    let mut mem = 0f64;
    let cur_in = repeat(0f64)
        .take(10)
        .chain(repeat(0.2f64).take(190))
        .collect::<Vec<_>>();
    let mut mem_rec = Vec::new();
    let mut spk_rec = Vec::new();
    let num_steps = 200;
    for step in 0..num_steps {
        let (new_mem, spk_out) = lif1.step(mem, cur_in[step]);
        mem = new_mem;
        mem_rec.push(mem);
        spk_rec.push(spk_out);
    }
    for (i, p) in mem_rec.iter().enumerate() {
        println!("{i}: {p:.8}");
    }
    for (i, p) in spk_rec.iter().enumerate() {
        println!("{i}: {p}");
    }
}
